components/script/dom/console.rs

```diff
@@ -55,14 +64,138 @@ where
     f()
 }
 
-fn console_messages(global: &GlobalScope, messages: &[DOMString], level: LogLevel) {
-    console_message(global, DOMString::from(messages.join(" ")), level)
+const MAX_LOG_DEPTH: usize = 10;
+
+#[allow(unsafe_code)]
+unsafe fn value_to_source(cx: *mut jsapi::JSContext, value: HandleValue) -> DOMString {
+    rooted!(in(cx) let mut js_string = std::ptr::null_mut::<jsapi::JSString>());
+    js_string.set(JS_ValueToSource(cx, value));
+    return jsstring_to_str(cx, *js_string);
+}
+
+#[allow(unsafe_code)]
+fn stringify_message(message: HandleValue) -> DOMString {
+    let cx = *GlobalScope::get_cx();
+    unsafe {
+        if message.is_string() {
+            jsstring_to_str(cx, message.to_string())
+        } else {
+            unsafe fn stringify_inner(
+                cx: *mut jsapi::JSContext,
+                value: HandleValue,
+                mut parents: Vec<u64>,
+            ) -> DOMString {
+                if parents.len() >= MAX_LOG_DEPTH {
+                    return DOMString::from("...");
+                }
+                let value_bits = value.asBits_;
+                if parents.contains(&value_bits) {
+                    return DOMString::from("[circular]");
+                }
+                parents.push(value_bits);
```

Comment:
If this is `undefined` does it still make sense to push the value bits onto the parents?

Reply:
I moved the push call to be right before `stringify_object_from_handle_value` (which is just an optimization, the value bits of non-objects were never being read).